File: frontend/types/investigation.ts
export type Severity = "Critical" | "High" | "Medium" | "Low" | "Info";

export interface Remediation {
  immediate_fix: string;
  verification_steps: string[];
  rollback_steps: string[];
}

export interface Finding {
  id: string;
  title: string;
  severity: Severity;
  category: string;
  affected_resource: string;
  namespace: string;
  resource_type: string;
  current_state: string;
  evidence: string[] | string;
  root_cause?: string | null;
  explanation?: string | null;
  investigation_commands: string[] | string;
  remediation?: Remediation | null;
  confidence?: number | null;
  confidence_reasoning?: string | string[] | null;
}

export interface ClusterHealthSummary {
  nodes_ready: string;
  nodes_total: number;
  pods_running: number;
  pods_failed: number;
  pods_pending: number;
  pods_total: number;
  deployments_healthy: number;
  deployments_degraded: number;
  services_missing_endpoints: number;
  critical_findings: number;
  high_findings: number;
  warning_findings: number;
}

export interface TimelineStep {
  step: number;
  action: string;
  status: string;
}

export interface Diagnosis {
  root_cause: string;
  explanation: string;
  fix: string;
  kubectl_command: string;
  confidence: number;
  prevention_recommendation?: string;
  confidence_reasoning?: string | string[] | null;
  cluster_healthy?: boolean;
  executive_summary?: string;
  cluster_health_score?: number;
  cluster_health_summary?: ClusterHealthSummary;
  investigation_timeline?: TimelineStep[];
  findings?: Finding[];
  validation_commands?: string[];
  remediation_steps?: string[];
  verification_steps?: string[];
  rollback_commands?: string[];
}

export interface ClusterInfo {
  name: string;
  cluster: string;
  server: string;
  is_current: boolean;
  is_gke: boolean;
}

export interface ClustersResponse {
  healthy: boolean;
  error?: string | null;
  kubeconfig_path: string;
  current_context: string;
  clusters: ClusterInfo[];
}

export interface InvestigationPayload {
  cluster_context?: string | null;
  nodes?: Record<string, unknown>;
  pods: Record<string, unknown>;
  logs: Record<string, unknown>;
  events: Record<string, unknown>;
  deployments: Record<string, unknown>;
  network: Record<string, unknown>;
  replicasets?: Record<string, unknown>;
  ingress?: Record<string, unknown>;
  timeline?: TimelineStep[];
}

export interface InvestigateResponse {
  status: string;
  cluster_context?: string | null;
  investigation: InvestigationPayload;
  diagnosis: Diagnosis;
  session_id?: string | null;
}

export interface InvestigationHistoryItem {
  id: string;
  user_id: string;
  session_id?: string | null;
  root_cause: string;
  namespace: string;
  confidence: number;
  status: string;
  diagnosis?: Diagnosis | null;
  created_at: string;
}

export interface ProgressStep {
  id: string;
  label: string;
  status: "pending" | "in_progress" | "completed";
}

export const INVESTIGATION_STEPS: ProgressStep[] = [
  { id: "context", label: "Loading Cluster Context", status: "pending" },
  { id: "connect", label: "Connecting to GKE Cluster", status: "pending" },
  { id: "nodes", label: "Checking Nodes", status: "pending" },
  { id: "pods", label: "Checking Pods", status: "pending" },
  { id: "logs", label: "Reading Logs", status: "pending" },
  { id: "events", label: "Analyzing Events", status: "pending" },
  { id: "deployments", label: "Inspecting Deployments", status: "pending" },
  { id: "network", label: "Checking Networking", status: "pending" },
  { id: "ai", label: "AI Reasoning", status: "pending" },
  { id: "complete", label: "Diagnosis Ready", status: "pending" },
];

File: backend/models/investigation.py
from typing import Any, Literal

from pydantic import BaseModel, Field

Severity = Literal["Critical", "High", "Medium", "Low", "Info"]


class ClusterInfo(BaseModel):
    name: str
    cluster: str
    server: str
    is_current: bool = False
    is_gke: bool = False


class ClustersResponse(BaseModel):
    healthy: bool
    error: str | None = None
    kubeconfig_path: str
    current_context: str
    clusters: list[ClusterInfo]


class Remediation(BaseModel):
    immediate_fix: str = ""
    verification_steps: list[str] = Field(default_factory=list)
    rollback_steps: list[str] = Field(default_factory=list)


class Finding(BaseModel):
    id: str
    title: str
    severity: Severity
    category: str
    affected_resource: str
    namespace: str
    resource_type: str
    current_state: str
    evidence: list[str] = Field(default_factory=list)
    root_cause: str = ""
    explanation: str = ""
    investigation_commands: list[str] = Field(default_factory=list)
    remediation: Remediation = Field(default_factory=Remediation)
    confidence: int = Field(default=0, ge=0, le=100)
    confidence_reasoning: str = ""


class ClusterHealthSummary(BaseModel):
    nodes_ready: str = ""
    nodes_total: int = 0
    pods_running: int = 0
    pods_failed: int = 0
    pods_pending: int = 0
    pods_total: int = 0
    deployments_healthy: int = 0
    deployments_degraded: int = 0
    services_missing_endpoints: int = 0
    critical_findings: int = 0
    high_findings: int = 0
    warning_findings: int = 0


class TimelineStep(BaseModel):
    step: int
    action: str
    status: str = "completed"


class InvestigationPayload(BaseModel):
    cluster_context: str | None = None
    nodes: dict[str, Any] = {}
    pods: dict[str, Any]
    logs: dict[str, Any]
    events: dict[str, Any]
    deployments: dict[str, Any]
    network: dict[str, Any]
    replicasets: dict[str, Any] = {}
    ingress: dict[str, Any] = {}
    timeline: list[TimelineStep] = Field(default_factory=list)


class Diagnosis(BaseModel):
    # Backward-compatible primary fields (first/most severe finding)
    root_cause: str
    explanation: str
    fix: str
    kubectl_command: str
    confidence: int = Field(ge=0, le=100)
    prevention_recommendation: str = ""
    confidence_reasoning: str = ""
    cluster_healthy: bool = False

    # SRE report extensions
    executive_summary: str = ""
    cluster_health_score: int = Field(default=100, ge=0, le=100)
    cluster_health_summary: ClusterHealthSummary = Field(default_factory=ClusterHealthSummary)
    investigation_timeline: list[TimelineStep] = Field(default_factory=list)
    findings: list[Finding] = Field(default_factory=list)
    validation_commands: list[str] = Field(default_factory=list)
    remediation_steps: list[str] = Field(default_factory=list)
    verification_steps: list[str] = Field(default_factory=list)
    rollback_commands: list[str] = Field(default_factory=list)


class InvestigateRequest(BaseModel):
    session_id: str | None = None
    cluster_context: str | None = None


class InvestigateResponse(BaseModel):
    status: str
    cluster_context: str | None = None
    investigation: InvestigationPayload
    diagnosis: Diagnosis
    session_id: str | None = None
